### ISSUE TITLE
Experimental WebAssembly Support?

### ISSUE THREAD

**Issue by GummyBear101010** [gosu/gosu#406]
Did somebody try to use Gosu by compiling Ruby code to WebAssemby by using LLVM and [Asm.js](https://en.wikipedia.org/wiki/Asm.js)?

**Comment by jlnr** [comment]
This has been requested many times but I'm not a web developer, so I don't know if any of Gosu's dependencies are already web-ready: Ruby, SDL 2, OpenAL...

Input welcome. Running Ruby/Gosu games in the browser would be nice.

**Comment by pmer** [comment]
A few years ago, SDL 1 applications could be compiled to Asm.js. SDL 2 is probably used nowadays. I bet there's an easy tutorial that shows how to do it, and from there we could explore if SDL's audio works also (I bet it does).

**Comment by Aerotune** [comment]
Emscripten compiles to webassembly and should have bindings to SDL2 and OpenAL.

**Comment by jlnr** [comment]
Thanks for the pointers! I am really tempted to try mruby again. They have a mode where the whole application is compiled to bytecode in a `.c` file, [shown here under "Bytecode (.c)"](http://mruby.org/docs/articles/executing-ruby-code-with-mruby.html). I *think* that could make it easier to use emscripten, but mostly it would also make it easy to compile (or even cross-compile?) Ruby games on all platforms, from iOS to Windows. The Ruby.app project has always felt like a huge burden :/

**Comment by shawn42** [comment]
This seems relevant as well: https://github.com/blacktm/ruby-wasm/

**Comment by jlnr** [comment]
And since we are already talking about mruby more than about "Ruby proper", I also have a permanent itch to try porting Gosu to Crystal once there's a Windows port: https://news.ycombinator.com/item?id=17281996

I usually write my games in a very, very basic subset of Ruby (class, def, if, collections) and Crystal generally looks like a better fit for game development. However, I'm not sure if it has enough momentum for anyone to port it to the browser :/

**Comment by Kjarrigan** [comment]
:+1: for a crystal port.

**Comment by bkmgit** [comment]
One other option is to use [xpra](https://xpra.org/), see in particular https://xpra.org/trac/wiki/Clients/HTML5

**Comment by rubyFeedback** [comment]
It would be nice to use as an additional subset. My dream is to write a DSL that can combine
different GUIs, such as ruby-gtk, gosu ... and the www too.

 jlnr pointed out lack of domain knowledge which I can understand. But I hope that gosu itself
can remain as flexible as possible, as a DSL. There are lots of different projects that are really
great (libui + the ruby bindings to it for example). The more flexible we can be, sooner or later
someone will add some relevant knowledge and make things work! \o/

**Comment by fguillen** [comment]
5 years later.. is there any news about this?

**Comment by jlnr** [comment]
Yes, in the sense that there is a Crystal version of Gosu: https://github.com/gosu/gosu.cr

But Crystal does not support wasm yet either: https://github.com/crystal-lang/crystal/issues/12002

So, not really...

**Comment by orangeSi** [comment]
Crystal had support ```hello world``` or  some simple code for webassembly at https://github.com/crystal-lang/crystal/pull/1087

**Comment by jlnr** [comment]
That does not look like the correct issue? But thanks for the general pointer, it looks like some progress is still being made: https://github.com/lbguilherme/crystal-web/issues/2

**Comment by jlnr** [comment]
This talk seems to go into some detail about how C extension linking works with ruby.wasm. Seems extremely relevant if for some reason we have to stop the FFI migration and go back to SWIG or something else altogether.
https://www.youtube.com/watch?v=_4d20A6xIsM&t=1268s

**Comment by jlnr** [comment]
And another link, WASM support in mruby also seems to be improving: https://github.com/mruby/mruby/pull/6487
